### PR TITLE
Add transformation code

### PR DESCRIPTION
**Pull request by seanbudd**
Adds the following:
Code to implement the transformation, which can be run as a module (see README)
Unit tests for the code and datastructures used

Tested with:
https://github.com/nvaccess/addon-store-submission/runs/3719448185?check_suite_focus=true#step:4:101
https://github.com/nvaccess/addon-store-submission/commit/0516c52ed72db7bbab0d89363c8072840e752221

### CODE REVIEW

File: src/transform/transform.py [hunk 1]
```diff
@@ -2,11 +2,135 @@
...
+		VersionCompatibility(
+			nvdaVersion=NVDAVersion.fromStr(version["NVDAVersion"]),
+			apiVer=NVDAVersion.fromStr(version["apiVer"]),
+			backCompatTo=NVDAVersion.fromStr(version["backCompatTo"]),
+		) for version in NVDAVersionData
+	)
+
+
+def emptyDirectory(dir: str) -> None:
+	for filename in glob.glob(dir + "/**/*.json", recursive=True):
```
Comment (feerrenrut) [inline src/transform/transform.py:122]: Shouldn't this delete all files, not just json files?
You could consider using `shutil.rmtree(dir)`

File: src/transform/datastructures.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	However, AddonVersion.fromStr("1.3.0") should preserve the patch number.
+	"""
+	major: int
+	minor: int
+	patch: Union[int, bool] = False  # is compared as 0, but works as a flag for hiding the patch by default
+
+	@classmethod
+	def fromStr(cls, version: str):
+		versionTuple = tuple(int(v) for v in version.split("."))
+		return MajorMinorPatch(*versionTuple)
```
Comment (feerrenrut) [inline src/transform/datastructures.py:24]: I think it would be good to be consistent with the version number scheme. Either `src/validate/addon_data.schema.json` should be updated to use the "canonical" version number used in the submissions schema, or the submissions schema should be updated to use a string.

Using 3 integers allows the json library to do the work of validation and separating the parts (at the expense of greater verbosity in the file). I suspect that these files will be created by a form on a GH issue eventually, so I'm not so worried about the verbosity.

File: src/transform/datastructures.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
+# Copyright (C) 2021 NV Access Limited
+# This file may be used under the terms of the GNU General Public License, version 2 or later.
+# For more details see: https://www.gnu.org/licenses/gpl-2.0.html
+
+from dataclasses import dataclass
+from typing import Dict, Literal, NamedTuple, Optional, OrderedDict, Union
+
+AddonChannels = Literal["beta", "stable"]
```
Comment (feerrenrut) [inline src/transform/datastructures.py:8]: This is cool, I didn't know about it.

How does it compare to using an enum?

File: src/tests/test_transform.py [hunk 4]
```diff
@@ -2,15 +2,115 @@
...
+		"""Confirm an addon is not compatible with an old API"""
+		addon = MockAddon()
+		addon.minNVDAVersion = V_2021_1
+		addon.lastTestedVersion = V_2021_2
+		self.assertFalse(_isAddonCompatible(addon, nvdaVersion2020_3))
+
+
+class Test_getLatestAddons(unittest.TestCase):
+	def test_beta_stable(self):
+		"""Ensure addons are added to the correct channels"""
```
Comment (feerrenrut) [inline src/tests/test_transform.py:56]: I'm not really sure what this test is showing. It the two versions have different values (other than stable / beta), but don't no versions exceed boundary conditions.

File: src/tests/test_transform.py [hunk 5]
```diff
@@ -2,15 +2,115 @@
...
+		newAddon = MockAddon()
+		newAddon.addonId = "foo"
+		newAddon.minNVDAVersion = V_2020_1
+		newAddon.lastTestedVersion = V_2020_3
+		newAddon.channel = "beta"
+		newAddon.addonVersionNumber = AddonVersion(0, 2)
+		oldAddon = MockAddon()
+		oldAddon.addonId = "foo"
+		oldAddon.minNVDAVersion = V_2020_1
+		oldAddon.lastTestedVersion = V_2020_2
```
Comment (feerrenrut) [inline src/tests/test_transform.py:85]: If this is about testing version number comparisons, keep the other values constant. Ideally `newAddon` would be a copy of `oldAddon`, and just update the version number.

EG
```
		NVDAVersions = (nvdaVersion2020_3) # only one version necessary to demonstrate 
		oldAddon = MockAddon(
			channel = "beta"
			addonId = "foo",
			addonVersionNumber = AddonVersion(0, 1),
			minNVDAVersion = V_2020_1,
			lastTestedVersion = V_2020_3,
		)
		newAddon = copy(oldAddon)
		# This version should supercede 0.3 for nvdaVersion2020_3
		newAddon.addonVersionNumber = AddonVersion(0, 2)
```

File: src/tests/test_transform.py [hunk 6]
```diff
@@ -2,15 +2,115 @@
...
-class Test_testExample(unittest.TestCase):
-	def test_example_true(self):
-		self.assertTrue(True)
 
-	def test_example_pass(self):
-		pass
+class Test_isAddonCompatible(unittest.TestCase):
+	def test_valid_with_api(self):
+		"""Confirm an addon is compatible with a fully tested API"""
+		addon = MockAddon()
```
Comment (feerrenrut) [inline src/tests/test_transform.py:27]: What is the benefit of using a `MockAddon` over the real datatype in this case? Will it raise if you access an unset variable?

Typically I would use a mock to supply a specific value, or to spy on values given to the mock, or methods called.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
I suspect that there are more edge cases to test for `Test_getLatestAddons`, but given the feedback I've already got, it might be best to address that first. 

I haven't reviewed the eversion number parsing yet, on the assumption that it is replaced with the alternate json schema.

File: src/transform/datastructures.py (hunk 3 above)
Comment (seanbudd) [reply src/transform/datastructures.py:8]: I believe it's a Python3.8+ feature - I think you can use it in 3.7 with `typing_extensions`.
It's only useful for type hinting, as opposed to the other features of an Enum.

I think the only argument in favour of using it is that it is less rigid than an Enum.
If we wanted to, we could make this project accept any string for channel.

All that would need to change:
* accept any string for `channel`
* remove `generateAddonChannelDict`, instead create the channel keys as needed.

"beta" and "stable" are only used in tests and in these instances. If we wanted to use an `Enum` we could replace the test instances with a reference to the Enum member.

File: src/transform/transform.py (hunk 1 above)
Comment (feerrenrut) [reply src/transform/transform.py:122]: I'd either expect it to delete all files, or make it the responsibility of the user to specify an empty / non-existent dir. I tend to agree with the safety for devs argument, but in that case check if the folder exists and exit with an error if it does.

Leaving files not generated from the source, could result in non reproducible outcome. The view should be entirely generated from the source. Keeping files means it has "state" of it's own.

File: src/tests/test_transform.py (hunk 6 above)
Comment (feerrenrut) [reply src/tests/test_transform.py:27]: Please expand on the comment in `def MockAddon() -> Addon:` to explain.

File: src/transform/datastructures.py (hunk 3 above)
Comment (feerrenrut) [reply src/transform/datastructures.py:8]: So does this mean that if there isn't runtime validation for the values allowed?

In this case, the validation is already done by the json schema.

Comment (seanbudd) [reply src/transform/datastructures.py:8]: Yes, exactly.
There is also run time validation in this instance, as only "beta" and "stable" channels are created in `generateAddonChannelDict`.
This could be changed to be handled by the schema only.

Comment (feerrenrut) [reply src/transform/datastructures.py:8]: Perhaps just add a comment to this literal to point out that the valid runtime values are defined by the json schema.

Comment (seanbudd) [reply src/transform/datastructures.py:9]: Done as of b40a61d

File: src/tests/test_transform.py (hunk 6 above)
Comment (seanbudd) [reply src/tests/test_transform.py:27]: Done as of b40a61d

File: src/transform/datastructures.py (hunk 2 above)
Comment (seanbudd) [reply src/transform/datastructures.py]: This was written before that was introduced and we had to rely on the file name. I've just moved this code to the NVDAVersion datastructure.

File: src/transform/transform.py (hunk 1 above)
Comment (seanbudd) [reply src/transform/transform.py]: Done as of 82cef85

File: src/tests/test_transform.py (hunk 4 above)
Comment (seanbudd) [reply src/tests/test_transform.py:58]: I've re-written this to make the purpose clearer.
I'm not sure what boundary conditions you mean?

File: src/tests/test_transform.py (hunk 5 above)
Comment (seanbudd) [reply src/tests/test_transform.py]: Good idea, I've improved this now.

File: src/transform/transform.py [hunk 7]
```diff
@@ -2,11 +2,129 @@
...
+
 
 def runTransformation(nvdaVersionsPath: str, sourceDir: str, outputDir: str) -> None:
 	"""
 	Performs the transformation of addon data described in the readme.
 	Takes addon data found in sourceDir that fits the schema and writes the transformed data to outputDir.
 	Uses the NVDA API Versions found in nvdaVersionsPath.
 	"""
-	pass
+	# Make sure the director doesn't already exist so data isn't overwritten
```
Comment (seanbudd) [inline src/transform/transform.py:126]:
```suggestion
	# Make sure the directory doesn't already exist so data isn't overwritten
```

File: src/tests/test_transform.py (hunk 4 above)
Comment (feerrenrut) [reply src/tests/test_transform.py:58]: Sorry, I'm not sure how I mangled that comment so much. I think I meant to point out that despite the different values, none of the different values cross boundaries which affect the behavior.

An example of testing either side of some boundary:
Setup two addon releases, older addon version that supports version 1 and 2 and a newer addon version that supports version 2. You could show that for version 1 == older addon release, for version 2 == newer addon

This code has a few of these boundaries. Look at the logic in the `_isAddonCompatible` and `_isAddonNewer` functions.

File: src/transform/transform.py [hunk 8]
```diff
@@ -2,11 +2,129 @@
...
+def readNVDAVersionInfo(pathToFile: str) -> Tuple[VersionCompatibility]:
+	"""
+	Reads and captures NVDA version information from file.
+	"""
+	with open(pathToFile, "r") as NVDAVersionFile:
+		NVDAVersionData = json.load(NVDAVersionFile)
+	validateJson(NVDAVersionData, JSONSchemaPaths.NVDA_VERSIONS)
+	return tuple(
+		VersionCompatibility(
+			nvdaVersion=NVDAVersion.fromStr(version["NVDAVersion"]),
```
Comment (feerrenrut) [inline src/transform/transform.py:113]: It doesn't look like `nvdaVersion` is used?

File: src/transform/datastructures.py (hunk 2 above)
Comment (feerrenrut) [reply src/transform/datastructures.py]: Still, I think either this and NVDAVersions.json should be updated to use 3 integers, or the other repositories should be updated to use strings and regex.

File: src/tests/test_datastructures.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+		self.assertLess(NVDAVersion(13, 2), NVDAVersion(13, 2, 1))
+		self.assertGreater(NVDAVersion(3, 2, 1), NVDAVersion(1, 2, 3))
+
+	def test_toStr(self):
+		"""Test converting versions to string"""
+		self.assertEqual(NVDAVersion(1, 3, 4).toStr(), "1.3.4")
+		self.assertEqual(NVDAVersion(13, 2).toStr(), "13.2")
+		self.assertEqual(NVDAVersion(13, 2, 0).toStr(), "13.2.0")
+
+	def test_fromStr(self):
```
Comment (feerrenrut) [inline src/tests/test_datastructures.py:26]: Using three integers for versions in the `nvdaVersion.json` file would allow deleting all the "fromStr" code and tests.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
I don't think `addon_data.schema.json` should be duplicated, it will make it harder to stay in sync. The data has to be validated on submission so should already be valid (and validation should be the responsibility of the validation repo).

We seem to have a mismatch for version numbers, in some cases two ints are allowed, in other places it must be three. This is likely to cause issues if there is a mismatch.

File: src/transform/transform.py (hunk 8 above)
Comment (seanbudd) [reply src/transform/transform.py]: I've removed it from the schema

File: src/tests/test_datastructures.py (hunk 9 above)
Comment (seanbudd) [reply src/tests/test_datastructures.py]: I've done this

**Comment by seanbudd** [comment]
>I don't think addon_data.schema.json should be duplicated, it will make it harder to stay in sync. The data has to be validated on submission so should already be valid (and validation should be the responsibility of the validation repo).

If the schema changes, the current code will fail. Currently this schema only defines what is required for the code to run successfully.
I think it's worth keeping to catch errors early and clearly document changes to the internal API

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Overall looks good.

There could also be a test for two addons with versions that match (shouldn't happen, the validation should catch this). Raise an error.

File: src/tests/test_transform.py [hunk 10]
```diff
@@ -72,17 +71,22 @@ def test_beta_stable(self):
...
+		oldAddon.addonVersion = MajorMinorPatch(0, 1)
+		newAddon = deepcopy(oldAddon)
+		newAddon.addonVersion = MajorMinorPatch(0, 2)
+		newAddon.minNVDAVersion = V_2020_3
+		newAddon.lastTestedVersion = V_2020_3
+		newAddon.pathToData = "new-path"
 		self.assertDictEqual(getLatestAddons([oldAddon, newAddon], NVDAVersions), {
-			V_2020_2: {"beta": {"foo": newAddon}, "stable": {}},
+			V_2020_2: {"stable": {"foo": oldAddon}, "beta": {}},
+			V_2020_3: {"stable": {"foo": newAddon}, "beta": {}},
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply src/tests/test_transform.py]: I've added some comments to demonstrate the test coverage, what do you think?

File: nvdaAPIVersions.json [hunk 11]
```diff
@@ -82,5 +89,18 @@
 			"minor": 1,
 			"patch": 0
 		}
+	},
+	{
+		"description": "NVDA 2021.2",
```
Comment (feerrenrut) [inline nvdaAPIVersions.json:94]: Good catch

File: src/tests/test_transform.py [hunk 12]
```diff
@@ -16,88 +16,140 @@
...
 		newAddon.pathToData = "new-path"
-		self.assertDictEqual(getLatestAddons([oldAddon, newAddon], NVDAVersions), {
+
+		self.assertDictEqual(getLatestAddons([oldAddon, newAddon], nvdaAPIVersions), {
+			# When adding 1st addon: oldAddon.addonId not in addons
+			# When adding 2nd addon: pass as incompatible
 			V_2020_2: {"stable": {"foo": oldAddon}, "beta": {}},
+			# When adding 1st addon: oldAddon.addonId not in addons
+			# When adding 2nd addon: newAddon.addonVersion > oldAddon.addonVersion
 			V_2020_3: {"stable": {"foo": newAddon}, "beta": {}},
```
Comment (feerrenrut) [inline src/tests/test_transform.py]: I don't really understand this comment, specifically the "adding 1st/2nd addon". This only needs to be framed with reference to `getLatestAddons`, unless this is explaining some general "why" for the system?

I've just noted that this is testing that the order of the addons doesn't affect the output.
To make that clearer you could put the 'expected' into a constant, so it's easy to confirm it is the same for both cases.
A comment to draw attention to the order of the addon list would also help.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: src/tests/test_transform.py (hunk 12 above)
Comment (seanbudd) [reply src/tests/test_transform.py]: I've updated the comments and documentation further to explain this

File: src/tests/test_transform.py [hunk 13]
```diff
@@ -2,15 +2,164 @@
...
+
+		newAddon = deepcopy(oldAddon)
+		newAddon.addonVersion = MajorMinorPatch(0, 3)
+		newAddon.minNvdaAPIVersion = V_2020_3
+		newAddon.lastTestedVersion = V_2020_3
+		newAddon.pathToData = "new-path"
+
+		self.assertDictEqual(getLatestAddons([oldAddon, newAddon], nvdaAPIVersions), {
+			# 1st addon is added because: oldAddon.addonId not in addons
+			# 2nd addon it not added: pass as incompatible
```
Comment (seanbudd) [inline src/tests/test_transform.py:101]:
```suggestion
			# 2nd addon is not added: pass as incompatible
```

File: src/tests/test_transform.py [hunk 14]
```diff
@@ -2,15 +2,164 @@
...
+			V_2020_2: {"stable": {"foo": oldAddon}, "beta": {}},
+			# 1st addon is added because: oldAddon.addonId not in addons
+			# 2nd addon updates it because: newAddon.addonVersion > oldAddon.addonVersion
+			V_2020_3: {"stable": {"foo": newAddon}, "beta": {}},
+		})
+
+		# Reverse order of addons
+		self.assertDictEqual(getLatestAddons([newAddon, oldAddon], nvdaAPIVersions), {
+			# 1st addon is not added because: pass as incompatible
```
Comment (seanbudd) [inline src/tests/test_transform.py:110]:
```suggestion
			# 1st addon is not added because: version is incompatible
```

File: src/tests/test_transform.py (hunk 12 above)
Comment (feerrenrut) [reply src/tests/test_transform.py]: Perhaps "included" would be clearer? Or can you answer: "added to what?"

Comment (seanbudd) [reply src/tests/test_transform.py]: maybe "stored"? as in they are stored in the addons dictionary.

Comment (feerrenrut) [reply src/tests/test_transform.py]: Oh, ok now I see what you're trying to say. It would be clearer to make it a full sentence. I'd discard the "obvious" bit (old.addonId not in addons), presumably that is not the behavior you wish to draw attention to here, and is covered in other tests.

EG
```
# For V_2020_2 newAddon is incompatible
...
# For V_2020_3 oldAddon overrides newAddon because newAddon.addonVersion > oldAddon.addonVersion
```

File: src/tests/test_transform.py [hunk 15]
```diff
@@ -70,8 +70,17 @@ def test_beta_stable(self):
...
+	def test_isAddonNewer(self):
+		"""
+		Ensure only the newest addon is used for a version+channel.
+
+		getLatestAddons checks addons in order, and we want to ensure
+		the addon dictionary is updated to maintain the newest addon
+		for a version+channel.
+
+		This tests _isAddonNewer by confirming that the order of addons being added
+		does not effect the results.
```
Comment (feerrenrut) [inline src/tests/test_transform.py:82]: > the order of addons being added

This doesn't fit the abstraction, it requires knowledge of the internals of `getLatestAddons`. For someone without that familiarity, it sounds like "added to the addon-store-submissions" but that doesn't make sense here either.

Instead, I'd suggest talking about it in terms of: "the order of addons in the list (`addons` arg) does not affect the output."

File: src/tests/test_transform.py (hunk 12 above)
Comment (seanbudd) [reply src/tests/test_transform.py]: I've turned this into a generator to make the ordering a bit more clear/important and fixed the comments.

File: src/tests/test_transform.py [hunk 16]
```diff
@@ -96,23 +95,17 @@ def test_isAddonNewer(self):
...
 		newAddon.pathToData = "new-path"
 
-		self.assertDictEqual(getLatestAddons([oldAddon, newAddon], nvdaAPIVersions), {
-			# 1st addon is added because: oldAddon.addonId not in addons
-			# 2nd addon it not added: pass as incompatible
-			V_2020_2: {"stable": {"foo": oldAddon}, "beta": {}},
-			# 1st addon is added because: oldAddon.addonId not in addons
-			# 2nd addon updates it because: newAddon.addonVersion > oldAddon.addonVersion
-			V_2020_3: {"stable": {"foo": newAddon}, "beta": {}},
+		addonGenerator = (addon for addon in (oldAddon, newAddon))
```
Comment (feerrenrut) [inline src/tests/test_transform.py:98]: I don't think the generator does anything to aid readability here, why use it over a list in this case?

File: src/tests/test_transform.py [hunk 17]
```diff
@@ -96,23 +95,17 @@ def test_isAddonNewer(self):
...
-			# 1st addon is not added because: pass as incompatible
-			# 2nd addon is added because: oldAddon.addonId not in addons
-			V_2020_2: {"stable": {"foo": oldAddon}, "beta": {}},
-			# 1st addon is added because: newAddon.addonId not in addons
-			# 2nd addon does not update it because: oldAddon.addonVersion < newAddon.addonVersion
-			V_2020_3: {"stable": {"foo": newAddon}, "beta": {}},
+		addonGenerator = (addon for addon in (newAddon, oldAddon))
+		self.assertDictEqual(getLatestAddons(addonGenerator, nvdaAPIVersions), {
+			# For V_2020_3 oldAddon does not override newAddon because newAddon.addonVersion > oldAddon.addonVersion
+			V_2020_3: {"stable": {"foo": newAddon}, "beta": {}}
```
Comment (feerrenrut) [inline src/tests/test_transform.py:108]: Since the goal of this test is to show that the output does not depend on the input, why not just compare them directly:
```python
# Ensure order of the addon list does not matter.
self.assertDictEqual(
	getLatestAddons([oldAddon, newAddon], nvdaAPIVersions),
	getLatestAddons([newAddon, oldAddon], nvdaAPIVersions)
)
# For ease of development, verify output
self.assertDictEqual(
	getLatestAddons([newAddon, oldAddon], nvdaAPIVersions),
	V_2020_3: {"stable": {"foo": newAddon}, "beta": {}}
)

File: src/tests/test_transform.py [hunk 18]
```diff
@@ -74,18 +74,17 @@ def test_isAddonNewer(self):
...
-		getLatestAddons checks addons in order, and we want to ensure
-		the addon dictionary is updated to maintain the newest addon
-		for a version+channel.
+		getLatestAddons checks addons in order, and we want to ensure the addon dictionary is updated to
+		maintain the newest addon for a version+channel.
 
-		This tests _isAddonNewer by confirming that the order of addons being added
-		does not effect the results.
+		This tests that the newest addons is always stored by confirming that the order of addons supplied
+		does not effect the output.
```
Comment (feerrenrut) [inline src/tests/test_transform.py:81]: Should there be a similar test to verify the order of the NVDA API Version data?

File: src/tests/test_transform.py (hunk 16 above)
Comment (seanbudd) [reply src/tests/test_transform.py:98]: It guarantees the order that the addons are consumed, and better replicates how the function is used internally.
I don't think it's super important though, as the function should support both datastructures as iterators.

File: src/tests/test_transform.py (hunk 17 above)
Comment (seanbudd) [reply src/tests/test_transform.py:108]: I don't think this says "why" the test is occurring, what the test aims to achieve. Why should the order of the addon list not matter? Because we want to assert that the newest addon is always used, and this exercises the two main pathways in `_isAddonNewer`.

File: src/tests/test_transform.py (hunk 18 above)
Comment (seanbudd) [reply src/tests/test_transform.py:81]: I don't really see what that test would involve or why the ordering of the NVDA API Version matters, as it's not used in any comparisons.

Comment (feerrenrut) [reply src/tests/test_transform.py:81]: Just thinking about the function stand-alone. The function takes two lists, and outputs a dictionary. [Dictionaries in python 3.7 are ordered](https://mail.python.org/pipermail/python-dev/2017-December/151283.html).

The things the caller can control are the contents of the two lists, and the order of the contents of the two lists. If for some reason the order of the API versions changes, does the output change?

File: src/tests/test_transform.py (hunk 17 above)
Comment (feerrenrut) [reply src/tests/test_transform.py:108]: Yes, you could add a comment to the example say, "output relies entirely on the relative properties of the addons, not their order". But I'd say that aspect is more obvious in one of the other tests.

File: src/tests/test_transform.py (hunk 16 above)
Comment (feerrenrut) [reply src/tests/test_transform.py:98]: Not really, the method could have it's first lines:
```
addons = list(addons)
addons.reverse()
```

File: src/tests/test_transform.py (hunk 18 above)
Comment (seanbudd) [reply src/tests/test_transform.py:81]: Does the order of the output dictionary matter? I would be surprised if `assertDictEqual` checks the ordering.

Comment (seanbudd) [reply src/tests/test_transform.py]: The following test passes:

```py
	def test_assertDictEq(self):
		d1 = {}
		d1["first"] = 1
		d1["second"] = 2
		d2 = {}
		d2["second"] = 2
		d2["first"] = 1
		self.assertDictEqual(d1, d2)
```

Comment (feerrenrut) [reply src/tests/test_transform.py]: Interesting, it probably should fail, given the order of the dict may now be something that is relied upon. I guess that's the danger with changing dicts to be ordered.

File: tox.ini [hunk 19]
```diff
@@ -21,4 +21,4 @@ commands =
 	python -m unittest discover src/tests
 	# lint will only run if tests pass.
 	flake8 --config=flake8.ini src
```
Comment (feerrenrut) [inline tox.ini:23]: This should run for tests too

Comment (seanbudd) [reply tox.ini:23]: `tests` are in `src`

File: src/tests/test_transform.py (hunk 18 above)
Comment (seanbudd) [reply src/tests/test_transform.py]: Is this something we need to rely upon?

File: tox.ini (hunk 19 above)
Comment (feerrenrut) [reply tox.ini:23]: Oh my mistake.

**Review by feerrenrut — APPROVED** [review]
Good work!

**Comment by seanbudd** [comment]
Thank you 🎉
